src/System.Threading.Channels/src/System/Threading/Channels/AsyncOperation.cs

```diff
@@ -0,0 +1,370 @@
+        protected static void ThrowIncorrectCurrentIdException() =>
+            throw new InvalidOperationException(SR.InvalidOperation_IncorrectToken);
+
+        /// <summary>Describes states the operation can be in.</summary>
+        public enum States
+        {
+            /// <summary>The operation has been assigned an owner.  No one else can use it.</summary>
+            Owned = 0,
+            /// <summary>Completion has been reserved.  Only the reserver is allowed to complete it.</summary>
+            CompletionReserved = 1,
+            /// <summary>The operation has completed and has had its result or error stored.</summary>
+            CompletionSet = 2,
+            /// <summary>The operation's result/error has been retrieved.  It's available for reuse.</summary>
+            Released = 3
+        }
+    }
+
+    internal abstract class ResettableValueTaskSource<T> : ResettableValueTaskSource, IValueTaskSource, IValueTaskSource<T>
+    {
+        private volatile int _state = (int)States.Owned;
+        private T _result;
+        private ExceptionDispatchInfo _error;
+        private Action<object> _continuation;
+        private object _continuationState;
+        private object _schedulingContext;
+        private ExecutionContext _executionContext;
+        private short _currentId;
+
+        public ValueTask CreateNonGenericValueTask() => new ValueTask(this, _currentId);
+        public ValueTask<T> CreateGenericValueTask() => new ValueTask<T>(this, _currentId);
+
+        public bool RunContinutationsAsynchronously { get; protected set; }
+        public ValueTaskSourceStatus GetStatus(short token)
+        {
+            if (_currentId != token)
+            {
+                throw new InvalidOperationException();
+            }
+
+            switch ((States)_state)
+            {
+                case States.Owned:
+                case States.CompletionReserved:
+                    return ValueTaskSourceStatus.Pending;
+
+                case States.CompletionSet:
+                case States.Released:
+                    return
+                        _error == null ? ValueTaskSourceStatus.Succeeded :
+                        _error.SourceException is OperationCanceledException ? ValueTaskSourceStatus.Canceled :
+                        ValueTaskSourceStatus.Faulted;
+
+                default:
+                    Debug.Fail($"Shouldn't be accessed in the '{(States)_state}' state.");
+                    goto case States.CompletionSet;
+            }
+        }
+
+        public bool IsCompleted => _state >= (int)States.CompletionSet;
+        public States UnsafeState { get => (States)_state; set => _state = (int)value; }
+
+        public T GetResult(short token)
+        {
+            if (_currentId != token)
+            {
+                ThrowIncorrectCurrentIdException();
+            }
+
+            if (!IsCompleted)
```

Comment:
This will allow GetResult to succeed if we're in the Released state, right?  Do we want to do that?

Reply:
Or does the _currentId check prevent this?  If so, should we assert we're not in released state?  Maybe we can't reliably do that.

Reply:
> Or does the _currentId check prevent this?

Right.

> If so, should we assert we're not in released state?

I try to avoid adding asserts for things that are actually triggerable using public APIs.  In this case such an assert would even be triggered by our tests that validate an exception is thrown when GetResult is called again.